src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h"
+
+#include <algorithm>
+#include <deque>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "src/common/json/json.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mqtt {
+
+// MatchKey layout, || control_packet_type (4 bits) | dup (1 bit) | qos (2 bits) | retain (1 bit) ||
+typedef uint8_t MatchKey;
+
+constexpr MatchKey UnmatchedResp = 0xff;
+
+std::map<MatchKey, MatchKey> MapRequestToResponse = {
+    // CONNECT to CONNACK
+    {0x10, 0x20},
+    // PUBLISH QOS 0 to Dummy response
+    {0x30, UnmatchedResp},
+    {0x31, UnmatchedResp},
+    {0x38, UnmatchedResp},
+    {0x39, UnmatchedResp},
+    // PUBLISH QOS 1 to PUBACK
+    {0x32, 0x40},
+    {0x33, 0x40},
+    {0x3a, 0x40},
+    {0x3b, 0x40},
+    // PUBLISH QOS 2 to PUBREC
+    {0x34, 0x50},
+    {0x35, 0x50},
+    {0x3c, 0x50},
+    {0x3d, 0x50},
+    // PUBREL to PUBCOMP
+    {0x60, 0x70},
+    // SUBSCRIBE to SUBACK
+    {0X80, 0X90},
+    // UNSUBSCRIBE to UNSUBACK
+    {0xa0, 0xb0},
+    // PINGREQ to PINGRESP
+    {0xc0, 0xd0},
+    // AUTH to AUTH
```

Comment:
I thought we decided that we were skipping supporting `AUTH` frames for now. If that changed, I'm ok with keeping this. However, I'd like to add a comment explaining what the gap is (if any) while it's still fresh in our minds.

Reply:
I decided to remove the mapping from the map itself, and a TODO before the logic of skipping AUTH matching in the stitcher function.